src/libbson/src/bson/bson-string.c

```diff
@@ -61,16 +62,23 @@ bson_string_t *
 bson_string_new (const char *str) /* IN */
 {
    bson_string_t *ret;
+   size_t len_sz;
 
    ret = bson_malloc0 (sizeof *ret);
-   ret->len = str ? (int) strlen (str) : 0;
+   if (str) {
+      len_sz = strlen (str);
+      BSON_ASSERT (len_sz <= INT32_MAX);
```

Comment:
On closer read, though bsonspec.org implies the limit of a string within BSON is `INT32_MAX`, I expect `bson_string_t` does not have the same limit. A `bson_string_t` is a string type within libbson, not necessary used within a BSON document.

Suggest assuming limit `UINT32_MAX` to match the type of `bson_string_t::len`:

```suggestion
      BSON_ASSERT (len_sz <= UINT32_MAX);
```

Reply:
I don't think we can do that. The `alloc` field is a `uint32_t`. If we allow a length anywhere from `INT32_MAX + 2` (2147483649) up to `UINT32_MAX` (4294967295), then the next power of two, 4294967296, will overflow the `uint32_t`. Setting the limit at `INT32_MAX` is the only way to ensure that the *next* power of 2 still fits in a `uint32_t`.

Reply:
Put more succinctly, the root of the problem is that we want to allocate memory in increments of powers of 2, but the integer a `uint32_t` can hold is not a power of 2 (i.e., `2^32 - 1`)

Reply:
OK. I ran some local tests, and it appears that I was only half right.

It seems like we can assert based on `UINT32_MAX` in all cases and in the event that the size of the string to create/append requires allocating all the way to 2^32, then the assertion fails.